include/sentry.h

```diff
@@ -477,32 +477,60 @@ SENTRY_API int sentry_envelope_write_to_file(
     const sentry_envelope_t *envelope, const char *path);
 
 /**
- * Type of the callback for transports.
+ * This represents an interface for user-defined transports.
+ *
+ * Transports are responsible for sending envelopes to sentry and are the last
+ * step in the event pipeline. A transport has the following hooks:
+ *
+ * * `send_func`: This function will take ownership of an envelope, and is
+ *   responsible for freeing it via `sentry_envelope_free`.
+ * * `startup_func`: This hook will be called by sentry and instructs the
+ *   transport to initialize itself.
+ * * `shutdown_func`: Instructs the transport to flush its queue and shut down.
+ *   This hook receives a millisecond-resolution `timeout` parameter and should
+ *   return `true` when the transport was flushed and shut down successfully.
+ *   In case of `false`, sentry will log an error, but continue with freeing the
+ *   transport.
+ * * `free_func`: Frees the provided `data`. This hook might be called even
+ *   though `shudown_func` returned `false` previously.
+ *
+ * The transport interface might be extended in the future with hooks to flush
+ * its internal queue without shutting down, and to dump its internal queue to
+ * disk in case of a crash.
```

Comment:
Should we also mention `data` here and what it should be used for? Also, I'm wondering if we can give this a different name, like `state`, that indicates the intent a bit better.

**Edit**: Looks like internally we call this `sentry_transport_state` already, so naming it `state` sounds good.